runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaMetricsBundleProgressHandler.java

```diff
@@ -0,0 +1,150 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.samza.runtime;
+
+import static org.apache.beam.runners.core.metrics.MonitoringInfoConstants.TypeUrns.DISTRIBUTION_INT64_TYPE;
+import static org.apache.beam.runners.core.metrics.MonitoringInfoConstants.TypeUrns.LATEST_INT64_TYPE;
+import static org.apache.beam.runners.core.metrics.MonitoringInfoConstants.TypeUrns.SUM_INT64_TYPE;
+import static org.apache.beam.runners.core.metrics.MonitoringInfoEncodings.decodeInt64Counter;
+import static org.apache.beam.runners.core.metrics.MonitoringInfoEncodings.decodeInt64Distribution;
+import static org.apache.beam.runners.core.metrics.MonitoringInfoEncodings.decodeInt64Gauge;
+
+import java.util.Map;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.model.pipeline.v1.MetricsApi;
+import org.apache.beam.runners.core.metrics.DistributionData;
+import org.apache.beam.runners.core.metrics.MonitoringInfoConstants;
+import org.apache.beam.runners.fnexecution.control.BundleProgressHandler;
+import org.apache.beam.runners.samza.metrics.SamzaMetricsContainer;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Distribution;
+import org.apache.beam.sdk.metrics.Gauge;
+import org.apache.beam.sdk.metrics.MetricName;
+import org.apache.beam.sdk.metrics.MetricsContainer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@inheritDoc} Parses metrics information contained in the bundle progress messages. Passed the
+ * updated metrics to the provided SamzaMetricsContainer.
+ */
+class SamzaMetricsBundleProgressHandler implements BundleProgressHandler {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SamzaMetricsBundleProgressHandler.class);
+  private final String stepName;
+
+  private final SamzaMetricsContainer samzaMetricsContainer;
+  private final Map<String, String> transformIdToUniqueName;
+
+  /**
+   * Constructor of a SamzaMetricsBundleProgressHandler.
+   *
+   * <p>The full metric names in classic mode is {transformUniqueName}:{className}:{metricName}. We
+   * attempt to follow the same format in portable mode, but the monitoringInfos returned by the
+   * worker only contains the transformId. The current solution is to provide a mapping from
+   * transformId back to uniqueName. A future improvement would be making the monitoring infos
+   * contain the uniqueName.
+   *
+   * @param stepName Default stepName provided by the runner.
+   * @param samzaMetricsContainer The destination for publishing the metrics.
+   * @param transformIdToUniqueName A mapping from transformId to uniqueName for pTransforms.
+   */
+  public SamzaMetricsBundleProgressHandler(
```

Comment:
suggest removing Metrics from the name since this isn't tied to metrics alone rather an implementation of `BundleProgressHandler` which can expand beyond metrics handling.

Reply:
Adding Metrics in the class name is pretty intentional. The name keeps the responsibility smaller. If we want a GeneralBundleProgressHandler, we can chain these smaller ones.

Reply:
I don't think BundleFactory takes multiple `BundleProgressHandler`s. The responsibility of `BundleProgressHandler` to listen to the signal of progress and take actions. It is agnostic to what actions are taken within and hence metric isn't part of the implementation.

If you want to take multiple actions within the handler, it could be delegated to multiple components but the one front facing component that bridges the actions and listening is done by this implementation.

Reply:
We can write something like
```
private List<BundleProgressHandler> handlers;
//Constructor
public GeneralBundleProgressHandler(List<BundleProgressHandler> handlers){
  this.handlers = handlers;
}

public void handle(ProcessBundleProgressResponse progress) {
  this.handlers.forEach(handler -> handler.handle(progress))
}
```

Reply:
I like delegating to multiple components. It's also a possible solution in the future. It's an overkill now since we only have 1 component.

I think the naming with Metric will prevent us from writing all code in 1 handler in the future, and make the developer think about either chaining or using components.

Reply:
You can't write something like that is what I was pointing out above. the signature of remote bundle is as follows

```  /** Get a new {@link RemoteBundle bundle} for processing the data in an executable stage. */
  default RemoteBundle getBundle(
      OutputReceiverFactory outputReceiverFactory,
      TimerReceiverFactory timerReceiverFactory,
      StateRequestHandler stateRequestHandler,
      BundleProgressHandler progressHandler)
      throws Exception {
    return getBundle(
        outputReceiverFactory, timerReceiverFactory, stateRequestHandler, progressHandler, null);
  }```

Given you can't have multiple handlers, name is naturally misleading to assume that there can be other handlers and that is the pattern to follow. `StageBundleFactory` doesn't seem to support and hence the suggestion to remove metrics.

Reply:
I meant writing the `GeneralBundleProgressHandler` as said above (Or rename as `ChainedBundleProgressHandler`), then you can chain them up by

```
BundleProgressHandler bundleProgressHandler = new ChainedBundleProgressHandler(List.of(samzaMetricsBundleProgressHandler, anotherBundleProgressHandler, yetAnotherBundleProgressHandler));

RemoteBundle remoteBundle = getBundle(..., bundleProgressHandler);
```

Reply:
The chained bundle handlers pattern is a bit hard to read and follow because you can have multiple of these handlers. 
As opposed to having one handler which delegates to appropriate components as part of handling.

I feel removing metrics from the name doesn't take away any of what its doing and keeps the door open regardless of both of the approaches.

Reply:
The Delegating handler pattern has the same problem. You will still inject the components in the constructor, unless you hard code them, which makes it less flexible.

Reply:
I think keeping Metric in the class name is a stronger name to push for refactoring, if we ever want to handler to do more stuff.

Reply:
@xinyuiscool @alnzng Please help make a final call.

Reply:
Let's go with the Metric in the name for now. I agree more with Bharath on this one: having a single general handler here might be easier to understand. Chaining might be an overkill. We can discuss this further and refactor in the future.